File: advent-of-code/day04/python/one.py
import os

input_path = os.path.join(os.path.dirname(__file__), "..", "input.txt")

def count_accessible() -> int:
    with open(input_path, "r") as file:
        input = file.readlines()

        rows = len(input)
        cols = len(input[0].strip())
        adjacent_directions = [(-1, -1), (-1, 0), (-1, 1),
                               (0, -1),          (0, 1),
                               (1, -1),  (1, 0),  (1, 1)]

        ans = 0
        for r in range(rows):
            for c in range(cols):
                if input[r].strip()[c] != "@":
                    continue
                adj = 0
                for dr, dc in adjacent_directions:
                    rr, cc = r + dr, c + dc
                    if 0 <= rr < rows and 0 <= cc < cols and input[rr].strip()[cc] == "@":
                        adj += 1
                if adj < 4:
                    ans += 1
        
        print(ans)
        return ans
    
count_accessible()

File: advent-of-code/day04/javascript/one.js
const fs = require("fs");
const path = require("path");

const inputPath = path.join(__dirname, "..", "input.txt");
const input = fs.readFileSync(inputPath, "utf-8").trim().split("\n");

function countAccessible(lines) {
  const rows = lines.length;
  const columns = rows ? lines[0].length : 0;

  const adjacentPositions = [
    [-1, -1],
    [-1, 0],
    [-1, 1],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 0],
    [1, 1],
  ];

  let ans = 0;
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < columns; c++) {
      if (lines[r][c] !== "@") continue;
      let adj = 0;
      for (const [moveRow, moveColumn] of adjacentPositions) {
        const rr = r + moveRow,
          cc = c + moveColumn;
        if (rr >= 0 && rr < rows && cc >= 0 && cc < columns && lines[rr][cc] === "@") {
          adj++;
        }
      }
      if (adj < 4) ans++;
    }
  }

  console.log("Total accessible '@' characters:", ans);
  return ans;
}

countAccessible(input);
